File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.java
public class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        if (s == null || s.length() == 0 || words == null || words.length == 0) {
            return result;
        }

        int wordLen = words[0].length();    // 单词的长度
        int wordCount = words.length;       // 单词的个数
        int totalLen = wordLen * wordCount; // 窗口长度
        
        // 构建目标单词频率表
        Map<String, Integer> wordMap = new HashMap<>();
        for (String word : words) {
            wordMap.put(word, wordMap.getOrDefault(word, 0) + 1);
        }

        // 从每个可能的起点开始滑动窗口
        for (int i = 0; i < wordLen; i++) {
            int left = i, right = i;  // 初始化窗口的左右边界
            int matchedCount = 0;     // 当前匹配的单词数量
            Map<String, Integer> currentMap = new HashMap<>();

            while (right + wordLen <= s.length()) {
                // 提取窗口内的单词
                String word = s.substring(right, right + wordLen);
                right += wordLen;

                // 如果单词在 words 中
                if (wordMap.containsKey(word)) {
                    currentMap.put(word, currentMap.getOrDefault(word, 0) + 1);

                    // 如果当前单词频率合法
                    if (currentMap.get(word) <= wordMap.get(word)) {
                        matchedCount++;
                    } else {
                        // 频率超出，缩小窗口
                        while (currentMap.get(word) > wordMap.get(word)) {
                            String leftWord = s.substring(left, left + wordLen);
                            left += wordLen;
                            currentMap.put(leftWord, currentMap.get(leftWord) - 1);

                            // 这里的 if 判断条件要弄清楚
                            // "<" 说明有效匹配的单词少一个 所以 macthedCount 要减少1
                            if (currentMap.get(leftWord) < wordMap.get(leftWord)) {
                                matchedCount--;
                            }
                        }
                    }

                    // 如果窗口匹配完整，记录起始索引
                    if (matchedCount == wordCount) {
                        result.add(left);
                    }
                } else {
                    // 如果单词不在 words 中，重置窗口
                    currentMap.clear();
                    matchedCount = 0;
                    left = right;
                }
            }
        }

        return result;
    }
}

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.cpp
// class Solution {
// public:
//     vector<int> findSubstring(string s, vector<string>& words) {
//         vector<int> ans;
//         if(words.empty() || s.empty()) return ans;

//         int wordLen = words[0].size();         // 计算一个word长度
//         int totalLen = wordLen * words.size(); // 计算总长度

//         unordered_map<string, int> wordFreq;
//         for(string& word : words) wordFreq[word]++;

//         // 枚举所有起点偏移: 0 到 wordLen - 1
//         for(int i = 0 ; i < wordLen ; i++){
//             int left = i, count = 0;
//             unordered_map<string, int> windowFreq;

//             for(int right = i ; right + wordLen <= s.size() ; right += wordLen){
//                 string word = s.substr(right, wordLen); // 截取单词

//                 // 合法词
//                 if(wordFreq.count(word)){
//                     windowFreq[word]++;
//                     count++;

//                     // 多了 缩小窗口
//                     while(windowFreq[word] > wordFreq[word]){
//                         string leftWord = s.substr(left, wordLen);
//                         windowFreq[leftWord]--;
//                         left += wordLen;
//                         count--;
//                     }

//                     if(count == words.size()) ans.push_back(left);

//                 } else {
//                     // 不合法词 清空窗口
//                     windowFreq.clear();
//                     count = 0;
//                     left = right + wordLen;
//                 }
//             }
//         }
//         return ans;
//     }
// };
// <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
 class Solution {
 public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> ans;
        if(s.empty() || words.empty()) return ans;

        unordered_map<string, int> wordFreq;
        for(string& word : words) wordFreq[word]++;

        int wordLen = words[0].size();
        int totalLen = wordLen * words.size();

        int left = 0;
        string window;
        unordered_map<string, int> windowFreq;
        for(int right = 0 ; right < s.size() ; right++){
            window += s[right];

            while(window.size() > totalLen){
                left++;
                window = window.substr(1);
            }

            if(window.size() == totalLen){
                for(int i = 0 ; i < window.size() ; i += wordLen){
                    string temp = window.substr(i, wordLen);
                    windowFreq[temp]++;
                }

                if(windowFreq == wordFreq){
                    ans.push_back(left);
                }

                windowFreq.clear();
            }
        }
        return ans;
    }
 };
